### PR TITLE
kvstore/etcd: fix etcd rate limit (QPS) not working

### PR DESCRIPTION
**Pull request by ArthurChiao**
kvstore/etcd sets a default QPS=20 for rate limiting, but this value is
accidently overriden by 0 (rateLimitOpt.value="" -> Atoi=0, and the error
is ignored), which eventually performs no rate limiting at all
(0 means Inf in `time/rate` package).

This patch fixes the problem, and also dumps the etcd cilent configuration
to log, which will faciliate trouble-shooting, as shown below:

```
level=info msg="Creating etcd client" ConfigPath=/var/lib/etcd-config/etcd.config KeepAliveHeartbeat=15s KeepAliveTimeout=25s RateLimit=30 subsys=kvstore
level=info msg="Creating etcd client" ConfigPath=/var/lib/cilium/clustermesh/cluster-1 KeepAliveHeartbeat=15s KeepAliveTimeout=25s RateLimit=20 subsys=kvstore
```

Example configuration:

* kvstore='etcd'
* kvstore-opt='{"etcd.config": "/var/lib/etcd-config/etcd.config", "etcd.qps": "30"}'

Signed-off-by: ArthurChiao <arthurchiao@hotmail.com>

### CODE REVIEW

**Review by tklauser — COMMENTED** [review]
Thanks, LGTM.

Could you please add the following to the commit message?

```
Fixes: b4e218089ebe ("Add rate limiting for etcd kvstore operations")
```

I think we also want to backport this fix, will add the respective labels.

File: pkg/kvstore/etcd.go [hunk 1]
```diff
@@ -187,20 +187,14 @@ type clientOptions struct {
...
 		KeepAliveTimeout:   25 * time.Second,
 		RateLimit:          defaults.KVstoreQPS,
 	}
 
-	if rateLimitSet {
-		// error is discarded here because this option has validation
-		clientOptions.RateLimit, _ = strconv.Atoi(rateLimitOpt.value)
+	if o, ok := e.opts[EtcdRateLimitOption]; ok && o.value != "" {
+		clientOptions.RateLimit, _ = strconv.Atoi(o.value)
 	}
```
Comment (aanm) [inline pkg/kvstore/etcd.go:198]: Wait a second, why is `rateLimitOpt` empty here? Wouldn't `rateLimitSet` be `false`? If it is `true` how is `rateLimitOpt` set to `""`?

**Review by aanm — CHANGES_REQUESTED** [review]

File: pkg/kvstore/etcd.go (hunk 1 above)
Comment (tklauser) [reply pkg/kvstore/etcd.go:198]: `rateLimitSet` (or `ok` in this case) will be `true` whenever the key `EtcdRateLimitOption` is present, regardless of the value stored in the map: https://play.golang.org/p/3YqsOD48lDK

Comment (ArthurChiao) [reply pkg/kvstore/etcd.go:198]: Hi @aanm , according to my understanding, `e.opts[EtcdRateLimitOption]` is initialized here: https://github.com/cilium/cilium/blob/master/pkg/kvstore/etcd.go#L135, with the `value` field unspecified. So when it comes to https://github.com/cilium/cilium/blob/master/pkg/kvstore/etcd.go#L193, 

```go
    rateLimitOpt, rateLimitSet := e.opts[EtcdRateLimitOption]
```

`rateLimitSet` will be `true` but `rateLimitOpt.value` will be `""`.

Comment (aanm) [reply pkg/kvstore/etcd.go:198]: @tklauser my question was more like how did the map end up having an empty value set to it. @ArthurChiao just explain it. Thanks!

Comment (aanm) [reply pkg/kvstore/etcd.go:198]: This also means that `endpointSet` and `configSet` have the same issue?

Comment (ArthurChiao) [reply pkg/kvstore/etcd.go:198]: > This also means that endpointSet and configSet have the same issue?

Yes, I think they are. But, for these two, the problem is hidden as they always use values from CLI inputs:

1.  kvstore.Setup() will pass the CLI parameters to kvstore package
2. `etcd.config` from CLI: defaults to `/var/lib/etcd-config/etcd.config`
2. `etcd.endpoints`: parsed from the contents of the config file `etcd.config` (e.g. `/var/lib/etcd-config/etcd.config`)

So the final values are correct.

@aanm

Comment (ArthurChiao) [reply pkg/kvstore/etcd.go:198]: Should I fix them in this patch as well?

Comment (aanm) [reply pkg/kvstore/etcd.go:198]: That would be great! It can be done as a separate PR.

**Review by aanm — APPROVED** [review]

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
hit https://github.com/cilium/cilium/issues/15620, merging

**Pull request merged**